### PR TITLE
[NEW] feat: Add sorting option for chatrooms

### LINKED ISSUE #880: [NEW] Support Sorting on the list of chats
Description:
Rocket Chat web and desktop allow for users to choose how to sort channels/DMs by activity/alphabetical/etc as well as how to group channels/DMs. The native Rocket Chat + Android/iOS mobile apps have no such control over sorting/grouping. This is highly desired by our users and impacts the way they communicate from their mobile devices.

Expected behavior:
Android/iOS mobile apps should at least replicate what users have set as their preference in the Desktop client, or allow them to configure sorting/grouping how they want.

Actual behavior:
Users have no control over sorting/grouping in the Rocket Chat + Android/iOS mobile apps.

I didn't include server details/logs as this isn't a bug or problem with the installation, but please let me know if they are required within the issue. Thanks!

### PR DESCRIPTION
**Pull request by Shailesh351**
@RocketChat/android

Closes #880 

**Changes**
- Added Sorting option in ChatRooms `Alphabetical`, `Activity`
- Added grouping by room type
- Saved user preference of sorting and grouping

**Checklist**
- [x] Add sorting
- [x] Add grouping by room type
- [x] Add headers for room type

**Feature task**
- Add grouping with favourites(Currently not available in sdk)

### CODE REVIEW

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -113,4 +113,10 @@
     <string name="status_disconnecting">disconnecting</string>
     <string name="status_waiting">connecting in %d seconds</string>
 
+    <string name="menu_chatroom_sort">Sort</string>
```
Comment (bizzbyster) [inline app/src/main/res/values/strings.xml]: I wonder if you should also provide at least "Google Translate" versions of these strings in the other supported languages. I'd think it'd be better than having missing strings, right?

Comment (Shailesh351) [reply app/src/main/res/values/strings.xml]: Sure @bizzbyster I will add translations after implementing complete feature :smile: Thanks!!!

**Comment by Shailesh351** [comment]
@rafaelks @bizzbyster @leonardoaramaki @filipedelimabrito @luciofm Ready to test and review :smile: 
 Added sorting and grouping. Works fine while searching also.

**Comment by Shailesh351** [comment]
**Demo**
demo

File: app/src/main/java/chat/rocket/android/chatrooms/ui/ChatRoomsFragment.kt [hunk 2]
```diff
@@ -147,15 +219,49 @@ class ChatRoomsFragment : Fragment(), ChatRoomsView {
...
+            var previousChatRoomType = ""
+
+            for ((position, chatRoom) in it.withIndex()) {
+                val type = chatRoom.type.toString()
+                if (type != previousChatRoomType) {
+                    val title = when (type) {
+                        RoomType.CHANNEL.toString() -> "Channels"
+                        RoomType.PRIVATE_GROUP.toString() -> "Private Groups"
+                        RoomType.DIRECT_MESSAGE.toString() -> "Direct Messages"
+                        RoomType.LIVECHAT.toString() -> "Live Chats"
```
Comment (bizzbyster) [inline app/src/main/java/chat/rocket/android/chatrooms/ui/ChatRoomsFragment.kt]: Pretty sure this doesn't support localization of strings (different languages or if an admin wants to rename "private groups" to "private rooms" but it sounds like you will do that in a later change. Just noting it here.

**Comment by bizzbyster** [comment]
@Shailesh351 I tried it out locally and all kinds of sorting worked for me. I'm new to Kotlin so my code review was very limited. Logic makes sense in any case. One newbie Rocket.chat question -- I would expect to have some tests for this new feature. Is there an Android test framework the team is using and do you think this feature should have a test or two covering basic functionality that can then be run by CircleCI?

File: app/src/main/java/chat/rocket/android/chatrooms/ui/ChatRoomsFragment.kt (hunk 2 above)
Comment (Shailesh351) [reply app/src/main/java/chat/rocket/android/chatrooms/ui/ChatRoomsFragment.kt]: Thanks @bizzbyster!!
I don't know. Is there any option in server to rename Headers?

File: app/src/main/res/values-pt-rBR/strings.xml [hunk 3]
```diff
@@ -122,4 +122,11 @@
     <string name="dialog_group_by_type">Agrupar por tipo</string>
     <string name="dialog_group_favourites">Grupos favoritos</string>
     <string name="chatroom_header">Cabeçalho</string>
+
+    <!--ChatRooms Headers-->
+    <string name="header_channel">Canais</string>
+    <string name="header_private_groups">"Grupos Privados "</string>
+    <string name="header_direct_messages">"Mensagens diretas "</string>
```
Comment (jaytat0) [inline app/src/main/res/values-pt-rBR/strings.xml]: @Shailesh351 are the quotes in these two lines correct? Should it just say Grupos Privados w/o quotes?

Comment (Shailesh351) [reply app/src/main/res/values-pt-rBR/strings.xml]: Right!! Thanks for reporting @jaytat0 Fixed it :smile:

**Comment by Shailesh351** [comment]
@filipedelimabrito @luciofm Can you have look on this? This will be very nice feature for 2.0.0

File: app/src/main/res/values-pt-rBR/strings.xml [hunk 4]
```diff
@@ -115,6 +115,22 @@
...
+    <string name="dialog_group_favourites">Grupos favoritos</string>
+    <string name="chatroom_header">Cabeçalho</string>
+
+    <!--ChatRooms Headers-->
+    <string name="header_channel">Canais</string>
+    <string name="header_private_groups">Grupos Privados</string>
+    <string name="header_direct_messages">Mensagens diretas</string>
+    <string name="header_live_chats">Live Chats</string>
+    <string name="header_unknown">Desconhecido</string>
+
```
Comment (leonardoaramaki) [inline app/src/main/res/values-pt-rBR/strings.xml]: :+1:

File: app/src/main/java/chat/rocket/android/app/RocketChatApplication.kt [hunk 5]
```diff
@@ -99,4 +102,11 @@ class RocketChatApplication : Application(), HasActivityInjector, HasServiceInje
     override fun serviceInjector(): AndroidInjector<Service> {
         return serviceDispatchingAndroidInjector
     }
+
+    companion object {
+        var context: WeakReference<Context>? = null
+        fun getAppContext(): Context? {
+            return context?.get()
+        }
+    }
```
Comment (leonardoaramaki) [inline app/src/main/java/chat/rocket/android/app/RocketChatApplication.kt:270]: I understand why you put this but we can use Dagger instead. I think will do for now :)

**Review by leonardoaramaki — APPROVED** [review]
Thanks for proactively adding PT-BR translations and seems like you tested many edge cases. Well done!

**Comment by Shailesh351** [comment]
@rafaelks Resolved conflicts. Have a look on this

**Comment by rafaelks** [comment]
Looking good to me! I'm gonna merge this one for now, since @leonardoaramaki already approved. 

@Shailesh351 what do you think about adding this other options in another PR? :-)

It's missing now:
- Group Favorites;
- Unread on top;

screen shot 2018-03-31 at 16 49 19

Later on, we can get the results from `/me` API and mirror the same setting from the web. 👍

**Comment by Shailesh351** [comment]
Sure @rafaelks!!  Will add grouping by favourites and unread in another PR :smile:
